### ISSUE TITLE
ZipArchive permissions on linux

### ISSUE THREAD

**Issue by chris-mackubin** [dotnet/runtime#17912]
ZipArchive doesn't currently allow manipulation permissions within.  I ran into this when creating file in memory, dropping it into a zip and handing off to another system.  With this code deployed on Linux, I'm not seeing a way to give the receiving system read access to contents of zip.

**Example:**

``` c#
    using (var memoryStream = new MemoryStream())
    {
        using (var archive = new ZipArchive(memoryStream, ZipArchiveMode.Create, true))
        {
            var lambdaFile = archive.CreateEntry("index.js");

            using (var entryStream = lambdaFile.Open())
            using (var streamWriter = new StreamWriter(entryStream))
            {
                streamWriter.Write(functionContent);
            }
        }

        memoryStream.Seek(0, SeekOrigin.Begin);
        svcFile.UploadFile(bucketName, s3Key, memoryStream);
    }
```

**Comment by daves0** [comment]
This is still an issue.  I'm creating a .zip archive in a MemoryStream using ZipArchive that is emailed to someone, and when the recipient opens the .zip and saves the contained file to disk, that file has NO read/write access permissions and cannot be used unless they change the file permissions using their OS.  Unlike Auspex above, I am not using a Windows server (I'm running on AWS Lambda) and this is a show-stopping bug for me.

**Comment by skaman** [comment]
I confirm that ZipArchive is not usable for create zip files on a linux server (for this problem with permissions). I had to use a third party library. I think it should generate readable zip files also on linux servers.

**Comment by marrrcin** [comment]
This is still valid issue, I don't understand why it's closed.

**Comment by danmosemsft** [comment]
Reopening in case IO area triage would like to see this feedback.

**Comment by jgimness** [comment]
I just ran into this problem today.

There's also issue  #1548 which looks related

**Comment by marcwittke** [comment]
```csharp
entry.ExternalAttributes = entry.ExternalAttributes | (Convert.ToInt32("664", 8) << 16);
```

...assuming `664` permissions

**Comment by ptsneves** [comment]
I confirm @marcwittke solution works. Thanks a lot man. This was quite a deal breaker

**Comment by v-yaremenko** [comment]
@marcwittke , @ptsneves , thanks for the provided solution.
However, I've faced that executable bit for user can't be set in a such way.
rw**X**rwxrwx <-- I mean the marked bit

My code:
`fileEntry.ExternalAttributes |= (Convert.ToInt32("100777", 8) << 16);`

Result (it is set to some default value):
` -rw-------     2.0 fat       24 b defN 20-Sep-18 17:35 fileA.txt`

But if I try to set 677, or some value without X bit for user, then it is okay:
`fileEntry.ExternalAttributes |= (Convert.ToInt32("100677", 8) << 16);`
` -rw-rwxrwx  2.0 fat       19 b defN 20-Sep-18 17:35 fileB.txt`

Would you please confirm that you are able to set the X bit for user?
If you have same issue, then it is a bug, as I understand.

**Comment by KirillOsenkov** [comment]
I think the bug is that the ExternalAttributes field on ZipArchiveEntry defaults to 0 and common entrypoint APIs don't set it:
https://github.com/dotnet/runtime/blob/7ab65423373f04998f3d365d564b3fc0e55fdf2b/src/libraries/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileExtensions.ZipArchive.Create.cs#L99

So if I'm using any standard high-level Zip creation API such as MSBuild's `ZipDirectory` task, which calls ZipFile.CreateFromDirectory (https://source.dot.net/#System.IO.Compression.ZipFile/System/IO/Compression/ZipFile.Create.cs,e30e0208967e9503), all the files will have permissions 0.

Combined with the global flag (I'm guessing) that controls whether the archive was created on Windows or not, the archive (when unzipped on Mac or Linux) will create files with 0 permissions.

We ran into this when we switched from building using Mono (which created archives that unpacked properly) to .NET.

**Comment by KirillOsenkov** [comment]
I think on Mac/Linux it should copy the permissions of the file being archived (just like it copies the timestamp). When creating on Windows, it should default to 644 for files (and maybe 755 for .exe files and scripts??)

**Comment by KirillOsenkov** [comment]
I think this might be a relevant link: https://stackoverflow.com/a/6297838/37899

**Comment by KirillOsenkov** [comment]
I'm using https://ide.kaitai.io to spelunk inside the binary format and I do notice that on a .zip file presumably created using Mono we have the upper byte of VersionMadeBy set to 0:
![image](https://user-images.githubusercontent.com/679326/110058566-6492e380-7d17-11eb-88aa-6db61961e868.png)

Whereas in the file that has 0 permissions that byte is set to 3:
![image](https://user-images.githubusercontent.com/679326/110058642-9015ce00-7d17-11eb-91a3-847e1ec64956.png)

which tells me that the unpacker doesn't default the permissions to 644 if the byte is set to 3:
https://source.dot.net/#System.IO.Compression/System/IO/Compression/ZipArchiveEntry.Unix.cs,0c3d2244301f9384,references

Haha and looking at the Mono implementation it has a bug:
https://github.com/mono/mono/blob/d035e181d1b640e3b40edd3c00d94874fd7245f5/mcs/class/System.IO.Compression/corefx/ZipArchiveEntry.Mono.cs#L7-L8

Basically Mono will always return "Windows" even when on Mac, because the `Path.PathSeparator` is `;`. They probably wanted `Path.DirectorySeparatorChar` instead ;)

**Comment by derwasp** [comment]
ZipArchiveEntry issue.

## Background

Recently we hit an issue where our customers saw that the artifact files in the zip archives that we produce have the permissions set to `000` if our system runs on Unix machines. When the `unzip` command is used, these permissions translate to the actual files also having the same permissions.  
This was an inconvenience to them as if the same system is ran on Windows, the files would have the `644` permissions. The customers wanted this to have the same behavior.

## Causes

Based on how I see it, the zip archives have two fields which can be responsible for the file permissions.  

### Version made by

One of them is `Version made by`. (See the `Central directory file header` section in the [wiki page](https://en.wikipedia.org/wiki/ZIP_(file_format)) ) This is what the comment says about this field in the code:

```
The upper byte of the "version made by" flag in the central directory header of a zip file represents the
OS of the system on which the zip was created. Any zip created with an OS byte not equal to Windows (0)
or Unix (3) will be treated as equal to the current OS.
The value of 0 more specifically corresponds to the FAT file system while NTFS is assigned a higher value. However
for historical and compatibility reasons, Windows is always assigned a 0 value regardless of file system.
```

So to simplify the story, we can say that we have two possible options here: `0` for `Windows` and `3` for `Unix`.

This field is written here: [ZipArchiveEntry.cs#L517](https://github.com/dotnet/runtime/blob/83adfae6a6273d8fb4c69554aa3b1cc7cbf01c71/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L517)

The unzipping software is using this field to determine that the permissions were present in the source system. In other words: we probably need to set the permissions to `644` on the Unix system if the archive was created on Windows, where these permissions don't exist. And this is exactly the behavior with `unzip` that we see.

### External attributes

Another field is `External attributes`. This field is used to store in the file permissions when the archives are created on the Unix systems. Recently, the API that works with files received support for this: [PR](https://github.com/dotnet/runtime/pull/55531).  
As you can see from the PR, the permissions are based on the actual permissions from the files:

```csharp
    static partial void SetExternalAttributes(FileStream fs, ZipArchiveEntry entry)
    {
        Interop.Sys.FileStatus status;
        Interop.CheckIo(Interop.Sys.FStat(fs.SafeFileHandle, out status), fs.Name);

        entry.ExternalAttributes |= status.Mode << 16;
    }
```

So now from the perspective of the unzipping software if we have a `Version made by` set to `3` (meaning `Unix`), we need to read the `External attributes` and apply the corresponding permissions to the files. And if `Version made by` is set to `0` (meaning `Windows`) we need determine the permissions ourself and probably set it to `644`.

### The issue

If the file API is used, then everything already works: the permissions are preserved and when the unpacking software is going to work with the archive it would create the files with same permissions as before.  
The caviat is that when we are creating a new `ZipArchiveEntry` from memory on the Unix machines. In this case, there's no file to begin with, but this new `ZipArchiveEntry` can still be unzipped using other software, such as `unzip`, which would follow the logic based on the `Version made by` flag.  

When this happens, currently, we would create an archive which, when unzipped, would result in files having `000` permissions if our zipping logic runs on `Windows` and `644` (probably depends on the implementation of the unzipping software) if our zipping logic runs on `Unix`.

Even if this is not techincally a bug, at least, this behavior is not consistent across the two platforms from the users point of view.

A minimal repro looks like this:

```csharp
using System.IO.Compression;

using var fileStream = new FileStream(@"test.zip", FileMode.Create);
using var archive = new ZipArchive(fileStream, ZipArchiveMode.Create, true);

ZipArchiveEntry demoFile = archive.CreateEntry("inmemory_file.txt");

using var entryStream = demoFile.Open();
using var streamWriter = new StreamWriter(entryStream);
streamWriter.Write("Hello zip!");

Console.WriteLine("Done.");

```

After running this on a `Unix` machine and unizipping the result with `unzip test.zip` you can do an `ls -la` and you would see something like this:

```
 xxxx@xxxx  ~/Projects/ConsoleApp1/ConsoleApp1/bin/Debug/net6.0  ls -la
drwxr-xr-x  10 derwasp  staff     320 Jun 14 14:48 .
drwxr-xr-x   3 derwasp  staff      96 Jun 13 16:42 ..
-rwxr-xr-x   1 derwasp  staff  132208 Jun 14 14:48 ConsoleApp1
-rw-r--r--   1 derwasp  staff     403 Jun 14 14:48 ConsoleApp1.deps.json
-rw-r--r--   1 derwasp  staff    5120 Jun 14 14:48 ConsoleApp1.dll
-rw-r--r--   1 derwasp  staff   10480 Jun 14 14:48 ConsoleApp1.pdb
-rw-r--r--   1 derwasp  staff     139 Jun 14 14:48 ConsoleApp1.runtimeconfig.json
----------   1 derwasp  staff      10 Jun 14 14:48 inmemory_file.txt
drwxr-xr-x   3 derwasp  staff      96 Jun 14 14:48 ref
-rw-r--r--   1 derwasp  staff     150 Jun 14 14:48 test.zip
```

And you can see that the file permissions are set to `000`.

### Workaround

To work this around one can do something like this:

```
    public static class ZipArchiveEntryExtensions
    {
        public static ZipArchiveEntry EnsureReadPermissions(this ZipArchiveEntry entry)
        {
            if (!OperatingSystem.IsWindows())
            {
                // BUGFIX
                // this is an equivalent of (Convert.ToInt32("644", 8) << 16)
                // where 644 are RW-R-R permissions
                const int readPermissions = 420 << 16;
                entry.ExternalAttributes |= readPermissions;
            }

            return entry;
        }
    }

```

And in the code, whenever you create a new entry, you would call the extension method like so (if we are modifying the repro from above):

```csharp
using System.IO.Compression;

using var fileStream = new FileStream(@"test.zip", FileMode.Create);
using var archive = new ZipArchive(fileStream, ZipArchiveMode.Create, true);

ZipArchiveEntry demoFile = archive.CreateEntry("inmemory_file.txt").EnsureReadPermissions();

using var entryStream = demoFile.Open();
using var streamWriter = new StreamWriter(entryStream);
streamWriter.Write("Hello zip!");

Console.WriteLine("Done.");
```

### Possible solution

It seems like the solution would be for us to set the permissions for such `ZipArchiveEntry`'s to be `644` at the time of creation. And it seems like we can do it in [ZipArchiveEntry.cs](https://github.com/dotnet/runtime/blob/83adfae6a6273d8fb4c69554aa3b1cc7cbf01c71/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs)'s constructors for the non-windows platforms. This, however, will change the current behavior and I am open to suggestions about which approach is best.

I would love to hear your thoughts on it @KirillOsenkov @danmoseley @ianhays

**Comment by danmoseley** [comment]
Cc @eerhardt who added the original change to preserve Unix permissions IIRC.

**Comment by eerhardt** [comment]
Has anyone looked at other implementations (Java, Python, Go, etc) to see what they do when creating .zip or .tar files from in-memory data? Do they have a hard-coded default permission (other than `000`)?

Looking at our own, new Tar implementation we seem to default to `644` as well:

https://github.com/dotnet/runtime/blob/a103efd28d46af39fc22a77458a11d204226e8d4/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs#L45

https://github.com/dotnet/runtime/blob/a103efd28d46af39fc22a77458a11d204226e8d4/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs#L22-L23

**Comment by derwasp** [comment]
I didn't look into the other implementations yet. Tar implementation looks promising though, thanks for bringing this up @eerhardt.

I made a [draft PR](https://github.com/dotnet/runtime/pull/70735) for the proposed fix so that we have it out there and we can collaborate further.
